### ISSUE TITLE
boot hang on kernels >= 4.14.36

### ISSUE THREAD

**Issue by kmjohansen** [linuxkit/linuxkit#3032]
Starting with kernel 4.14.36, I can no longer reach the login prompt when booting an image with docker and getty.  The kernel reaches init, but gets stuck trying to start containerd.

I've been testing this using the kernel+initrd build target, and then loading those images under qemu on both a local kvm instance, and one in the cloud.  It fails to reach login in both cases.

After putting getty into the init section, I'm able to look at containerd.
It has a thread blocked in the kernel with this stack:
```
[<0000000000000000>] wait_for_random_bytes+0x8c/0xbb
[<0000000000000000>] SyS_getrandom+0x68/0x83
[<0000000000000000>] do_syscall_64+0x74/0x87
[<0000000000000000>] entry_SYSCALL_64_after_hwframe+0x42/0xb7
[<0000000000000000>] 0xffffffffffffffff
```
Which corresponds to the following goroutine:
```
goroutine 29 [syscall, 1 minutes]:
syscall.Syscall(0x13e, 0xc4201e20e9, 0x3, 0x0, 0x3, 0x3, 0x0)
	/usr/lib/go/src/syscall/asm_linux_amd64.s:18 +0x5
internal/syscall/unix.GetRandom(0xc4201e20e9, 0x3, 0x3, 0x0, 0x0, 0x0, 0x0)
	/usr/lib/go/src/internal/syscall/unix/getrandom_linux.go:38 +0x6f
crypto/rand.getRandomLinux(0xc4201e20e9, 0x3, 0x3, 0x7fa9190cd9e3)
	/usr/lib/go/src/crypto/rand/rand_linux.go:23 +0x4a
crypto/rand.(*devReader).Read(0xc420074180, 0xc4201e20e9, 0x3, 0x3, 0x0, 0x0, 0x0)
	/usr/lib/go/src/crypto/rand/rand_unix.go:48 +0x3d3
io.ReadAtLeast(0x7fa91a072fe0, 0xc420074180, 0xc4201e20e9, 0x3, 0x3, 0x3, 0xc42020c380, 0x7fa919a3dd90, 0x0)
	/usr/lib/go/src/io/io.go:309 +0x88
io.ReadFull(0x7fa91a072fe0, 0xc420074180, 0xc4201e20e9, 0x3, 0x3, 0x7fa919baa5c0, 0xf8c2d01, 0xc4201e20e9)
	/usr/lib/go/src/io/io.go:327 +0x5a
crypto/rand.Read(0xc4201e20e9, 0x3, 0x3, 0xffc5e8bfa08ef1b9, 0xc42004d7f8, 0x3)
	/usr/lib/go/src/crypto/rand/rand.go:23 +0x59
github.com/containerd/containerd/services/leases.generateLeaseID(0xc42004d818, 0xc42004d728)
	/go/src/github.com/containerd/containerd/services/leases/service.go:113 +0x88
github.com/containerd/containerd/services/leases.(*service).Create(0xc42019e090, 0x7fa919941200, 0xc4201d2a80, 0xc4201e03a0, 0xc42019e090, 0x7fa9197f6b80, 0x6)
	/go/src/github.com/containerd/containerd/services/leases/service.go:55 +0x298
github.com/containerd/containerd/api/services/leases/v1._Leases_Create_Handler.func1(0x7fa919941200, 0xc4201d2a80, 0x7fa919c581e0, 0xc4201e03a0, 0xc4201d07d0, 0x7fa91a0de618, 0x7fa919b9d300, 0xc4201dc1c0)
	/go/src/github.com/containerd/containerd/api/services/leases/v1/leases.pb.go:208 +0x88
github.com/containerd/containerd/vendor/github.com/grpc-ecosystem/go-grpc-prometheus.UnaryServerInterceptor(0x7fa919941200, 0xc4201d2a80, 0x7fa919c581e0, 0xc4201e03a0, 0xc4201e03c0, 0xc4201e03e0, 0x50, 0x48, 0xc4201e0380, 0xc42004d968)
	/go/src/github.com/containerd/containerd/vendor/github.com/grpc-ecosystem/go-grpc-prometheus/server.go:29 +0xd4
github.com/containerd/containerd/server.interceptor(0x7fa919941200, 0xc4201d29c0, 0x7fa919c581e0, 0xc4201e03a0, 0xc4201e03c0, 0xc4201e03e0, 0x0, 0xc42004d9e0, 0x7fa9190ce14a, 0x50)
	/go/src/github.com/containerd/containerd/server/server.go:267 +0x1d2
github.com/containerd/containerd/api/services/leases/v1._Leases_Create_Handler(0x7fa919c0dd60, 0xc42019e090, 0x7fa919941200, 0xc4203a1980, 0xc4201d06e0, 0x7fa919cc5e40, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/containerd/containerd/api/services/leases/v1/leases.pb.go:210 +0x16f
github.com/containerd/containerd/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4201f8500, 0x7fa91a0837e0, 0xc4203fe160, 0xc4201ae800, 0xc420297110, 0x7fa91a064c60, 0x0, 0x0, 0x0)
	/go/src/github.com/containerd/containerd/vendor/google.golang.org/grpc/server.go:826 +0xab6
github.com/containerd/containerd/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4201f8500, 0x7fa91a0837e0, 0xc4203fe160, 0xc4201ae800, 0x0)
	/go/src/github.com/containerd/containerd/vendor/google.golang.org/grpc/server.go:1023 +0x152a
github.com/containerd/containerd/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4201e2130, 0xc4201f8500, 0x7fa91a0837e0, 0xc4203fe160, 0xc4201ae800)
	/go/src/github.com/containerd/containerd/vendor/google.golang.org/grpc/server.go:572 +0xa1
created by github.com/containerd/containerd/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/containerd/containerd/vendor/google.golang.org/grpc/server.go:570 +0xa3
```
The problem is this bit of code:
```
    SYSCALL_DEFINE3(getrandom, char __user *, buf, size_t, count,
            unsigned int, flags)
    {
        int ret;
    
        if (flags & ~(GRND_NONBLOCK|GRND_RANDOM))
            return -EINVAL;
    
        if (count > INT_MAX)
            count = INT_MAX;
    
        if (flags & GRND_RANDOM)
            return _random_read(flags & GRND_NONBLOCK, buf, count);
    
        if (!crng_ready()) {
            if (flags & GRND_NONBLOCK)
                return -EAGAIN;
 !# ===>    ret = wait_for_random_bytes();
            if (unlikely(ret))
                return ret;
        }
        return urandom_read(NULL, buf, count, NULL);
    }
```
The `crng_ready() `check is returning false, and since the getrandom syscall was not invoked with `GRND_NONBLOCK`, we're waiting for the crng to become ready.  Bummer.

The problem is caused by this patch, which arrived as a security fix in 4.14.36: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/commit/?h=v4.14.40&id=6e513bc20ca63f594632eca4e1968791240b8f18

It changes the `crng_ready()` check not to return true until the CRNG is _fully_ initialized, which means that in our low-entropy early boot stage we're blocking until we have enough entropy to proceed.

As a workaround, I added a oneshot call to the rngd to the onboot section:
```
  - name: rng-oneshot
    image: linuxkit/rngd:429e1308b8cad9dbe04b7a91fcebec17ee6f7591
    command: ["/sbin/rngd", "-1"]
```
This adds enough entropy to the pool during startup such that containerd no longer gets stuck trying to `generateLeaseID()`.  This is enough to get me to login.

**Comment by rn** [comment]
which platform are you booting this on. We run a whole bunch of boots in CI with QEMU and I hadn't had issue with 4.14.36 on hyperkit and a RPi3 either.

**Comment by kmjohansen** [comment]
This is qemu kvm on intel x86_64 and OEL7.

**Comment by rn** [comment]
Hm, I can boot the docker example with 4.14.39 on qemu 2.11.1 on a Ubuntu 18.04 system (4.15.0-20 kernel). The `rgnd` services exited with:
```
2018/05/10 07:47:45 No random source available
```

**Comment by ijc** [comment]
There was a regression in recent kernels to do with non-blocking `getrandom` calls, see https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=897572 and the merged bugs for example.

**Comment by justincormack** [comment]
qemu will often give no random sources available, thats not an issue. We dont currently run a one shot rngd in the examples, we probably ought to have one though, in case you want to do anything using randomness in init, and you do have random sources it can use. qemu should have virtio-rng wired up but if the kernel is messing up something could block on that.

**Comment by justincormack** [comment]
`rngd` has a oneshot mode `rngd -1` which may help in onboot if you have a source of entropy eg amd64 with `RDRAND`, this may allow the machine to boot sanely. You can try adding this to the start of `onboot`:

```
  - name: rngd1
    image: linuxkit/rngd:v0.3
    command: ["rngd", "-1"]
```

**Comment by kmjohansen** [comment]
@justincormack Thanks for the suggestion, and apologies if I was unclear before.  Adding a `rngd` oneshot to the `onboot` section was exactly what I did to get the machine to boot, once I realized that `containerd` was getting stuck waiting for `crng_ready()` to return true.  Wanted to get this on your radar in case adding this to the recipes was more generally applicable than to just my environment.

**Comment by kmjohansen** [comment]
@rn I'm sorry I don't have more clarity about what specifically in my environment makes this more likely to occur than in the one that you've been testing in.  I tested on qemu 2.0.0 and 2.9.0.  The host kernel was a 4.1.12 derivative used in UEK.   It's what's currently being used on the HVs in our cloud too.

**Comment by justincormack** [comment]
@kmjohansen ah sorry, I didnt read that above I think I missed the start of the thread my problem! It is generally applicable if starting up requires entropy, and given containerd does that will always be the case (I had ignored containerd when thinking about this before, network startup etc rarely require any). It certainly does no harm and we could usefully add it to some/all of the examples, and it could cause issues.

Are you running qemu with virtio rng support? `linuxkit run qemu` does
```
-object rng -device virtio-rng-pci,rng=rng0
```
which I would think would add enough, but the bug may well be ignoring it.

**Comment by kmjohansen** [comment]
@justincormack Thanks for that pointer.  I had not been running qemu with a `virtio-rng-pci` device.  Adding that to the mix does indeed cause boot to succeed without the need for a oneshot call to rngd.  It may be that neither the prod nor test environments are including this device by default.

**Comment by justinbarrick** [comment]
I ran into this issue despite having `virtio-rng-pci` and one-shot rngd in my onboot. I was able to resolve it loading `virtio-rng` with modprobe.

**Comment by justincormack** [comment]
Note CoreOS has decided to remove all requirements for entropy in early boot https://github.com/coreos/bugs/issues/2433 - this somewhat helps, and util-linux has an update that might be relevant (and might be in Alpine 3.8).

There are lkml threads; the kernel developers mostly seem to assume that you have some entropy source and your computer is going to have interactive logins, even though they do not enable most hardware sources (tpm, RDRAND etc) unless the user does something in userspace... eg see https://marc.info/?l=linux-kernel&m=152478701604805&w=2

The argument is that the previous config was broken and it has a CVE https://nvd.nist.gov/vuln/detail/CVE-2018-1108 so the behaviour was masked before. I think having an entropy source is probably a good idea at this point...

**Comment by justincormack** [comment]
Yes Alpine 3.8 does have the new util-linux, so a few operations should be better in master.

**Comment by justincormack** [comment]
Unfortunately `containerd` startup requires reading randomness, not sure about the rest of our early startup process.

**Comment by justincormack** [comment]
Yeah thats trying to work around it, but doesn't realise that the failure mode is blocking not returning nothing. The Linux `getrandom` syscall has a `GRND_NONBLOCK` flag that will return if there is nothing available, but Go does not use that.

**Comment by hairyhenderson** [comment]
@justincormack interestingly there is an internal `GetRandom` at https://golang.org/src/internal/syscall/unix/getrandom_linux.go that allows setting `GRND_NONBLOCK`, but I don't see that being referenced anywhere in the Go code.

In the lkml thread you [referenced](https://marc.info/?l=linux-kernel&m=152478701604805&w=2), the message from kernel devs seems to be "you might not need crypto-grade randomness immediately at boot time".

This begs the question - does containerd _need_ `crypto/rand` or could it make do with `math/rand`? And, perhaps an issue should be logged with containerd?

**Comment by hairyhenderson** [comment]
> And, perhaps an issue should be logged with containerd?

I've done that - containerd/containerd#2451

**Comment by justincormack** [comment]
Fedora have added the jitter entropy daemon see https://bugzilla.redhat.com/show_bug.cgi?id=1572944 this uses high resolution timers as a microarchitectural entropy source.

**Comment by justincormack** [comment]
@scriptjs we already do that if you do `linuxkit run ` see https://github.com/linuxkit/linuxkit/blob/master/src/cmd/linuxkit/run_qemu.go#L570 - if you are running yourself then I would recommend seeding the rng from `/dev/urandom` too.

**Comment by scriptjs** [comment]
kk, stuck on moving beyond 4.14.35 on openstack atm.

**Comment by scriptjs** [comment]
@justincormack This could help me. Will report on any success.
https://wiki.openstack.org/wiki/LibvirtVirtioRng#Random_number_generator_device

**Comment by rn** [comment]
Is this still an issue? I can boot images with newer kernels without issue on Linux

**Comment by kmjohansen** [comment]
I'll attempt to reproduce again this week, if I can find the time.  I've been using the rng oneshot at boot since I filed this ticket, but it would be nice to remove that if it's no longer needed.

**Comment by kmjohansen** [comment]
This one still remains problematic in my environment.  I removed the `rngd-oneshot` that I've got in my onboot section and was right back where I started with regard to not being able to get through the services section of my boot sequence.
